### DISCUSSION TITLE
Mass import of numerical values (equates)

### DISCUSSION THREAD

**Discussion by 0xBEEEF** [NationalSecurityAgency/ghidra#3415] [Q&A] [answered]
I am currently facing a problem. In a program that I am currently analyzing there are a lot of numerical values. From most of them I also know an assignment. In my case these represent strings. But they are not enumerations in the classical sense. In Ghidra I saw that there are equates for this. With this it should be possible to assign a meaning to a number.

What I have now: I have a list for the numbers and their meaning. I would have to get this now somehow in Ghidra. There I would have to replace all affected numbers with the numbers in the list. Unfortunately I can't find a script or something similar that supports me. There is an example about equates, but it doesn't really help me. And then, unfortunately, I can't find any proper documentation.

Therefore the question to the community:
Has anyone already built any ready-made scripts or tools to create masses of equates based on a file?

Thanks in advance!

**Comment by astrelsky** [comment, marked as answer]
If these are uncommon numbers that would only appear for some equate there is an option to apply to all occurances when you apply it to one in the listing.

**Reply by 0xBEEEF** [reply]
Yes, they are uncommon numbers. I have also seen the option with a single number. But in this case we are talking about numbers in a special number range. And there are hundreds if not thousands of numbers/equates. That means, if I do it all manually, I'll spend several months on it alone. So I thought that there is a simple way to 1. import the numbers including description from a CSV file and 2. automatically make all numbers occurring there to an Equate. Also, if it comes times to one or more wrong assignments thereby would not be so bad. The mass is in this case rather the problem than the false positives.

**Reply by astrelsky** [reply]
> Yes, they are uncommon numbers. I have also seen the option with a single number. But in this case we are talking about numbers in a special number range. And there are hundreds if not thousands of numbers/equates. That means, if I do it all manually, I'll spend several months on it alone. So I thought that there is a simple way to 1. import the numbers including description from a CSV file and 2. automatically make all numbers occurring there to an Equate. Also, if it comes times to one or more wrong assignments thereby would not be so bad. The mass is in this case rather the problem than the false positives.

You should be able to find the implementation that applies them all by searching the source for the option text. You should be able to use it from a script. If you can't find it I'll hunt it down after work.

**Reply by astrelsky** [reply]
@0xBEEEF here you go. `curScalar` is a `Scalar` holding the numerical value for the equate. This should be very easy to write a script to deserialize the enums you want from a format of your choice and apply them this way.

https://github.com/NationalSecurityAgency/ghidra/blob/8b2ea61e27c07c48dc21eff9095905f739208703/Ghidra/Features/Base/src/main/java/ghidra/app/plugin/core/equate/EquatePlugin.java#L212-L227

**Reply by 0xBEEEF** [reply]
Thanks for the tip on this. That is indeed enough for me to knit something myself.
